File: python/open_guardrail/guards/markdown_link_safety.py
"""Validate markdown links for security issues."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_MD_LINK_RE = re.compile(r"\[([^\]]*)\]\(([^)]+)\)")


class _MarkdownLinkSafety:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "markdown-link-safety"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        for m in _MD_LINK_RE.finditer(text):
            link_text = m.group(1)
            href = m.group(2).strip()

            if re.match(r"^javascript:", href, re.I):
                issues.append("javascript_uri")
            if re.match(r"^data:", href, re.I):
                issues.append("data_uri")
            if "../" in href or "..\\" in href:
                issues.append("path_traversal")
            if re.match(
                r"^https?://\d{1,3}(\.\d{1,3}){3}",
                href,
                re.I,
            ):
                issues.append("ip_based_link")
            if link_text.strip() == "":
                issues.append("hidden_link")

        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="markdown-link-safety",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=(
                min(len(issues) / 3, 1.0)
                if triggered
                else 0.0
            ),
            message=(
                "Unsafe markdown link detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def markdown_link_safety(
    *, action: str = "block"
) -> _MarkdownLinkSafety:
    return _MarkdownLinkSafety(action=action)

File: packages/guards/src/markdown-link-safety.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface MarkdownLinkSafetyOptions {
  action: 'block' | 'warn';
}

const MD_LINK_RE = /\[([^\]]*)\]\(([^)]+)\)/g;

export function markdownLinkSafety(options: MarkdownLinkSafetyOptions): Guard {
  return {
    name: 'markdown-link-safety',
    version: '0.1.0',
    description: 'Validates markdown links for javascript:, data: URIs, path traversal, and hidden links',
    category: 'security',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      let match: RegExpExecArray | null;
      const re = new RegExp(MD_LINK_RE.source, MD_LINK_RE.flags);
      while ((match = re.exec(text)) !== null) {
        const linkText = match[1];
        const href = match[2].trim();

        if (/^javascript:/i.test(href)) {
          issues.push('javascript_uri');
        }
        if (/^data:/i.test(href)) {
          issues.push('data_uri');
        }
        if (/\.\.\//.test(href) || /\.\.\\/.test(href)) {
          issues.push('path_traversal');
        }
        if (/^https?:\/\/\d{1,3}(\.\d{1,3}){3}/i.test(href)) {
          issues.push('ip_based_link');
        }
        if (linkText.trim() === '') {
          issues.push('hidden_link');
        }
      }

      const triggered = issues.length > 0;
      return {
        guardName: 'markdown-link-safety',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(issues.length / 3, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
